migration/identity-migration/src/main/java/io/camunda/migration/identity/handler/sm/TenantMigrationHandler.java

```diff
@@ -0,0 +1,238 @@
+    this.tenantService = tenantService.withAuthentication(camundaAuthentication);
+  }
+
+  @Override
+  protected List<Tenant> fetchBatch(final int page) {
+    return List.of();
+  }
+
+  @Override
+  protected void process(final List<Tenant> batch) {
+    final List<Tenant> tenants;
+    try {
+      tenants =
+          managementIdentityClient.fetchTenants().stream()
+              .filter(tenant -> !Objects.equals(tenant.tenantId(), DEFAULT_TENANT_IDENTIFIER))
+              .toList();
+    } catch (final Exception e) {
+      if (!isNotImplementedError(e)) {
+        throw new MigrationException("Failed to fetch tenants", e);
+      }
+      logger.warn("Tenant endpoint is not available, skipping.");
+      return;
+    }
+    totalTenantCount.set(tenants.size());
+
+    tenants.forEach(
+        tenant -> {
+          final var tenantId = normalizeID(tenant.tenantId());
+          try {
+            tenantService.createTenant(new TenantDTO(null, tenantId, tenant.name(), null)).join();
+            createdTenantCount.incrementAndGet();
+            logger.debug(
+                "Tenant with ID '{}' and name '{}' created successfully.", tenantId, tenant.name());
+          } catch (final Exception e) {
+            if (!isConflictError(e)) {
+              throw new MigrationException(
+                  "Failed to migrate tenant with ID: " + tenant.tenantId(), e);
+            }
+            logger.debug(
+                "Tenant with ID '{}' already exists, skipping creation.", tenant.tenantId());
+          }
+          assignUsersToTenant(tenant.tenantId(), tenantId);
+          assignGroupsToTenant(tenant.tenantId(), tenantId);
+          assignClientsToTenant(tenant.tenantId(), tenantId);
+        });
+  }
+
+  @Override
+  protected void logSummary() {
+    logger.info(
+        "Tenant migration completed: Created {} out of {} tenants, the remaining existed already. Assigned {} users out of {} attempted, the remaining were already assigned. Assigned {} groups out of {} attempted, the remaining were already assigned. Assigned {} clients out of {} attempted, the remaining were already assigned.",
+        createdTenantCount.get(),
+        totalTenantCount.get(),
+        assignedUserCount.get(),
+        totalUserAssignmentAttempts.get(),
+        assignedGroupCount.get(),
+        totalGroupAssignmentAttempts.get(),
+        assignedClientCount.get(),
+        totalClientAssignmentAttempts.get());
+  }
+
+  private void assignUsersToTenant(final String tenantId, final String normalizedTenantId) {
+    final List<User> tenantUsers;
+    try {
+      tenantUsers = managementIdentityClient.fetchTenantUsers(tenantId);
+    } catch (final Exception e) {
+      if (!isNotImplementedError(e)) {
+        throw new MigrationException("Failed to fetch users for tenant with ID: " + tenantId, e);
+      }
+      logger.warn("User tenant endpoint is not available, skipping user assignment.");
```

Comment:
❓ is there any scenario where this is possible? From my understanding either all or none of the tenant endpoints are there?

Reply:
Yeah, your understanding is correct. I still prefer to keep the try catch on every endpoint even if we are pretty sure that this is not possible